crates/virtio-queue/src/lib.rs

```diff
@@ -398,20 +400,143 @@ impl VirtqUsedElem {
 
 unsafe impl ByteValued for VirtqUsedElem {}
 
+/// Struct to hold an exclusive reference to the underlying `QueueState` object.
+pub enum QueueStateGuard<'a, M: GuestAddressSpace> {
+    /// A reference to a `QueueState` object.
+    StateObject(&'a mut QueueState<M>),
+    /// A `MutexGuard` for a `QueueState` object.
+    MutexGuard(MutexGuard<'a, QueueState<M>>),
+}
+
+impl<'a, M: GuestAddressSpace> Deref for QueueStateGuard<'a, M> {
+    type Target = QueueState<M>;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            QueueStateGuard::StateObject(v) => v,
+            QueueStateGuard::MutexGuard(v) => v.deref(),
+        }
+    }
+}
+
+impl<'a, M: GuestAddressSpace> DerefMut for QueueStateGuard<'a, M> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        match self {
+            QueueStateGuard::StateObject(v) => v,
+            QueueStateGuard::MutexGuard(v) => v.deref_mut(),
+        }
+    }
+}
+
+/// Trait to access and manipulate a virtio queue.
+///
+/// To optimize for performance, different implementations of the `QueueStateT` trait may be
+/// provided for single-threaded context and multi-threaded context.
+pub trait QueueStateT<M: GuestAddressSpace> {
+    /// Construct an empty virtio queue state object with the given `max_size`.
+    fn new(max_size: u16) -> Self;
+
+    /// Check whether the queue configuration is valid.
+    fn is_valid(&self, mem: &M::T) -> bool;
+
+    /// Reset the queue to the initial state.
+    fn reset(&mut self);
+
+    /// Get an exclusive reference to the underlying `QueueState` object.
+    ///
+    /// Logically this method will acquire the underlying lock protecting the `QueueState` Object.
+    /// The lock will be released when the returned object gets dropped.
+    fn lock(&mut self) -> QueueStateGuard<'_, M>;
```

Comment:
My first trial is to implement `fn lock(&mut self) -> impl DerefMut<Target=QueueState>` but it's really to hard to deal with the lifetime parameter needed. The I used the ugly `QueueStateGuard` to simplify the implementation.

The total goal of `lock()` method is to support following pattern:
```
fn process(queue: &mut Queue, eventfd: EventFd) -> Result<()> {
      loop {
          eventfd.read().unwrap();
          // Lock the queue, get the next descriptor chain, and unlock
          let chain = queue.lock().iter().next();
          ...
         // Lock the queue, add used descriptor, unlock.
          queue.add_used(chain.head_index);
      }     
}
```

Reply:
Ok so after thinking a bit in circles about this :D, it looks like the `lock` method ultimately adds to the interface some semantics that should probably be handled behind the scenes by specific implementations. For example, given that `QueueStateSync` is added in this PR as well (and it also implements `QueueStateT`), don't we get proper synchronized behavior implicitly whenever we're (or a `Queue` is) using that as the underlying state, without the need for an explicit `lock` method?

Reply:
That's the real art part:)
We are trying to balance between flexibility and performance. If we support multi-threading by simple replace single-threaded `QueueState` with multi-threaded version`QueueStateSync`, we may not get too much because `QueueStateSync` causes too many lock/unlock operations. So we want fine-granularity control of the lock policy. For casual users, it may just replace ` QueueState` with `QueueStateSync`, and everything should work as expected. For advanced users, it may call QueueStateSync::lock() to get the underlying guard object, and then optimize its lock policy.

Reply:
You're definitely right that we should keep in mind that different consumers might need different policies with varying degrees of granularity. In my mind this can be achieved by having different backends for `QueueStateT`, instead of also adding `lock` to the `QueueStateT` interface explicitly. If, at some point, a new use case appears that cannot be efficiently addressed via either `QueueState` or `QueueStateSync` (it looks like they're enough for now, please correct if I'm wrong), then a new backend can be plugged in (and also upstreamed if not super-specific to a particular project). Do you think this more incremental approach works based on what we know so far?

Reply:
We already have some clients needs the flexibility. The virtiofsd project and the nydus image service project also help to optimize performance by multi-threading. Our first trial has achieved some performance improvements, but it's still limited by the vm-queue interfaces. So I proposed this change to alleviate it.

Reply:
Seems we have two possible way to build the interface here.
One way is to let the virtio driver claims that "I'm a superman to support multi-threading concurrence, but you must supply a special queue with multi-threading capability to me, otherwise I will refuse to work".
Another way is that the virtio driver is optimized for multi-threading. When a multi-threading capable queue is passed in, the driver works in multi-threading mode. When a single-threaded queue is passed in, the driver works too.
The current interface is designed for the latter case.

Also it may also affect the implementation of device manager in vmm, I need more time to figure it out:)